### PR TITLE
feat(backend): Add login endpoint and controller functions (#13)

### PR DESCRIPTION
**Pull request by Rolly007**

### CODE REVIEW

File: backend/package.json [hunk 1]
```diff
@@ -21,10 +22,14 @@
     "http-status": "^1.3.1",
     "joi": "^14.3.0",
     "jsonwebtoken": "^8.5.1",
+    "lodash": "^4.17.15",
+    "moment": "^2.24.0",
```
Comment (revodark) [inline backend/package.json]: I think is not necessary anymore

File: backend/src/controllers/login.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    const appToken = await getGoMovieRequestToken();
+    const userData = { ...body, request_token: appToken };
+    const userValidatedToken = await loginTheMovieDB(userData);
+    const userSession = await generateUserSession({
+      request_token: userValidatedToken
+    });
+    const encryptedSessionId = encryptSessionId(userSession);
+    res.status(200).send(generateCokkie(encryptedSessionId));
+  } catch (err) {
+    const loginError = {
```
Comment (revodark) [inline backend/src/controllers/login.js]: Maybe we should control the 500 error too

File: backend/src/modules/services/movies.js [hunk 3]
```diff
@@ -15,4 +15,26 @@ const getMoviesByCategory = (category, sortBy) => {
...
+  const requestPath = `https://api.themoviedb.org/3/authentication/token/validate_with_login?api_key=${apiKey}`;
+  return Axios.post(requestPath, body).then(({ data }) => data.request_token);
+}
+
+function generateUserSession(requestToken) {
+  const requestPath = `https://api.themoviedb.org/3/authentication/session/new?api_key=${apiKey}`;
+  return Axios.post(requestPath, requestToken).then(
+    ({ data }) => data.session_id
+  );
+}
```
Comment (revodark) [inline backend/src/modules/services/movies.js]: I think that this should be a new file, maybe services/users

File: backend/src/controllers/login.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import cookie from 'cookie';
+
+import {
+  getGoMovieRequestToken,
+  loginTheMovieDB,
+  generateUserSession
+} from 'modules/services/login';
+import NodeRSA from 'node-rsa';
```
Comment (Rodri77) [inline backend/src/controllers/login.js]: Fix imports order

File: backend/src/controllers/login.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+const EXPIRE_COOKIE = 172800;
+
+const encryptKey = config.get('ENCRYPT_KEY');
+const key = new NodeRSA(encryptKey);
+
+const encryptSessionId = sessionId => {
+  return key.encrypt(sessionId, config.get('FORMAT_ENCRYPT'));
+};
+
+const generateCokkie = sessionId =>
```
Comment (Rodri77) [inline backend/src/controllers/login.js]: Fix your typo here for Cookie

File: backend/src/controllers/login.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    httpOnly: true,
+    maxAge: EXPIRE_COOKIE
+  });
+
+async function loginUser(req, res, next) {
+  const { body } = req;
+  try {
+    const appToken = await getGoMovieRequestToken();
+    const userData = { ...body, request_token: appToken };
+    const userValidatedToken = await loginTheMovieDB(userData);
```
Comment (Rodri77) [inline backend/src/controllers/login.js]: Don't we need to do some extra validation here before we generate user session? are we sure that userValidatedToken is always returned?

File: backend/src/controllers/login.test.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
+import AxiosTheMovieDB from 'axios';
+import NodeRSA from 'node-rsa';
```
Comment (Rodri77) [inline backend/src/controllers/login.test.js]: Don't forget about the import blocks and orders we agreed as conventions.

File: backend/src/controllers/login.test.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  const res = mockResponse('');
+  const next = () => {};
+
+  afterEach(() => {
+    jest.restoreAllMocks();
+  });
+
+  test('Test the LoginUser function without user', async () => {
+    const requestToken = '1531f1a558c8357ce8990cf887ff196e8f5402ec';
+    const validatedToken = '1531f1a558c8357ce8990cf887ff196e8f5402ec';
```
Comment (Rodri77) [inline backend/src/controllers/login.test.js]: Any reason why we need requestToken and validatedToken to have the same values? seems like we only need one here

File: backend/src/models/user.js [hunk 9]
```diff
@@ -1,19 +1,19 @@
...
-    name: { type: String, maxlength: 255, trim: true },
-    nickname: {
+    username: {
       type: String,
       maxlength: 100,
       trim: true,
       unique: true
     },
-    password: { type: String, required: true, minlength: 8, select: false }
+    password: { type: String, required: true, minlength: 8 },
```
Comment (Rodri77) [inline backend/src/models/user.js]: I think Pablo mentioned that we will only store the username in our DB, please double-check with him about this part.

File: backend/src/modules/services/login.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+import config from 'modules/system/config';
+
+const apiKey = config.get('API_KEY');
+
+const AxiosTheMovieDB = Axios.create({
+  baseURL: config.get('TMDB_BASE_URL')
+});
+
+function getGoMovieRequestToken() {
+  const requestPath = `3/authentication/token/new?api_key=${apiKey}`;
```
Comment (Rodri77) [inline backend/src/modules/services/login.js]: I think @revodark already has an env variable for this "3", if we don't have that yet we should create one to represent the api version we use, right now looks like a magic number.

File: backend/src/modules/services/login.test.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  });
+
+  test('Test getGoMovieRequestToken function ', async () => {
+    const requestToken = '1531f1a558c8357ce8990cf887ff196e8f5402ec';
+    AxiosTheMovieDB.get.mockImplementation(() =>
+      Promise.resolve({ data: { requestToken } })
+    );
+    const token = await AxiosRequest.getGoMovieRequestToken();
+    expect(AxiosTheMovieDB.get).toHaveBeenCalledTimes(1);
+    expect(token).toBe(token);
```
Comment (Rodri77) [inline backend/src/modules/services/login.test.js]: This test seems like a false positive because you are comparing the same variable. We need to fix this

**Pull request merged**